### PR TITLE
TxnMapProxy Made Generic

### LINKED ISSUE #7484: TransactionalMapProxy should be made generic
Given that ClientTxnMapProxy has generic types, it makes sense for TransactionalMapProxy to do so to. This will allow it to better support new method calls where the return type is a collection of keys and values, such as the getAll method proposed in #4698

### PR DESCRIPTION
**Pull request by aditya-32**
Making TransactionalMapProxy generic
Fixes #7484 

Checklist:
- [ ] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [ ] Label `Add to Release Notes` or `Not Release Notes content` set
- [ ] Request reviewers if possible
- [ ] Send backports/forwardports if fix needs to be applied to past/future releases
- [ ] New public APIs have `@Nonnull/@Nullable` annotations
- [ ] New public APIs have `@since` tags in Javadoc

### CODE REVIEW

**Comment by aditya-32** [comment]
hi @ahmetmircik @srknzl can you please review the PR

File: hazelcast/src/main/java/com/hazelcast/map/impl/tx/TransactionalMapProxy.java [hunk 1]
```diff
@@ -53,10 +53,10 @@
 /**
  * Proxy implementation of {@link TransactionalMap} interface.
  */
-public class TransactionalMapProxy
-        extends TransactionalMapProxySupport implements TransactionalMap {
+public class TransactionalMapProxy<K, V>
```
Comment (srknzl) [inline hazelcast/src/main/java/com/hazelcast/map/impl/tx/TransactionalMapProxy.java:56]: Changes I suggest for this file: 

```diff
diff --git a/hazelcast/src/main/java/com/hazelcast/map/impl/tx/TransactionalMapProxy.java b/hazelcast/src/main/java/com/hazelcast/map/impl/tx/TransactionalMapProxy.java
index 067409728b..9e4391dc32 100644
--- a/hazelcast/src/main/java/com/hazelcast/map/impl/tx/TransactionalMapProxy.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/tx/TransactionalMapProxy.java
@@ -140,12 +140,12 @@ public class TransactionalMapProxy<K, V>
     }
 
     @Override
-    public V put(Object key, Object value) {
+    public V put(K key, V value) {
         return put(key, value, -1, MILLISECONDS);
     }
 
     @Override
-    public V put(Object key, Object value, long ttl, TimeUnit timeUnit) {
+    public V put(K key, V value, long ttl, TimeUnit timeUnit) {
         checkTransactionState();
         checkNotNull(key, "key can't be null");
         checkNotNull(value, "value can't be null");
@@ -160,13 +160,13 @@ public class TransactionalMapProxy<K, V>
                 ttl, timeUnit, remoteCallHook));
         TxnValueWrapper<V> currentValue = txMap.get(keyData);
         Type type = valueBeforeTxn == null ? Type.NEW : Type.UPDATED;
-        TxnValueWrapper<V> wrapper = new TxnValueWrapper<>((V) value, type);
+        TxnValueWrapper<V> wrapper = new TxnValueWrapper<>(value, type);
         txMap.put(keyData, wrapper);
         return currentValue == null ? valueBeforeTxn : checkIfRemoved(currentValue);
     }
 
     @Override
-    public void set(Object key, Object value) {
+    public void set(K key, V value) {
         checkTransactionState();
         checkNotNull(key, "key can't be null");
         checkNotNull(value, "value can't be null");
@@ -179,12 +179,12 @@ public class TransactionalMapProxy<K, V>
 
         Data dataBeforeTxn = putInternal(keyData, valueData, UNSET, MILLISECONDS, remoteCallHook);
         Type type = dataBeforeTxn == null ? Type.NEW : Type.UPDATED;
-        TxnValueWrapper<V> wrapper = new TxnValueWrapper<>((V) value, type);
+        TxnValueWrapper<V> wrapper = new TxnValueWrapper<>(value, type);
         txMap.put(keyData, wrapper);
     }
 
     @Override
-    public V putIfAbsent(Object key, Object value) {
+    public V putIfAbsent(K key, V value) {
         checkTransactionState();
         checkNotNull(key, "key can't be null");
         checkNotNull(value, "value can't be null");
@@ -202,19 +202,19 @@ public class TransactionalMapProxy<K, V>
                 return wrapper.value;
             }
             putInternal(keyData, valueData, UNSET, MILLISECONDS, remoteCallHook);
-            txMap.put(keyData, new TxnValueWrapper<>((V) value, Type.NEW));
+            txMap.put(keyData, new TxnValueWrapper<>(value, Type.NEW));
             return null;
         } else {
             Data oldValue = putIfAbsentInternal(keyData, valueData, remoteCallHook);
             if (oldValue == null) {
-                txMap.put(keyData, new TxnValueWrapper<>((V) value, Type.NEW));
+                txMap.put(keyData, new TxnValueWrapper<>(value, Type.NEW));
             }
             return (V) toObjectIfNeeded(oldValue);
         }
     }
 
     @Override
-    public V replace(Object key, Object value) {
+    public V replace(K key, V value) {
         checkTransactionState();
         checkNotNull(key, "key can't be null");
         checkNotNull(value, "value can't be null");
@@ -232,19 +232,19 @@ public class TransactionalMapProxy<K, V>
                 return null;
             }
             putInternal(keyData, valueData, UNSET, MILLISECONDS, remoteCallHook);
-            txMap.put(keyData, new TxnValueWrapper<>((V) value, Type.UPDATED));
+            txMap.put(keyData, new TxnValueWrapper<>(value, Type.UPDATED));
             return wrapper.value;
         } else {
             Data oldValue = replaceInternal(keyData, valueData, remoteCallHook);
             if (oldValue != null) {
-                txMap.put(keyData, new TxnValueWrapper<>((V) value, Type.UPDATED));
+                txMap.put(keyData, new TxnValueWrapper<>(value, Type.UPDATED));
             }
             return (V) toObjectIfNeeded(oldValue);
         }
     }
 
     @Override
-    public boolean replace(Object key, Object oldValue, Object newValue) {
+    public boolean replace(K key, V oldValue, V newValue) {
         checkTransactionState();
         checkNotNull(key, "key can't be null");
         checkNotNull(oldValue, "oldValue can't be null");
@@ -269,7 +269,7 @@ public class TransactionalMapProxy<K, V>
             boolean success = replaceIfSameInternal(keyData, mapServiceContext.toData(oldValue),
                     newValueData, remoteCallHook);
             if (success) {
-                txMap.put(keyData, new TxnValueWrapper<>((V) newValue, Type.UPDATED));
+                txMap.put(keyData, new TxnValueWrapper<>(newValue, Type.UPDATED));
             }
             return success;
         }
@@ -403,7 +403,7 @@ public class TransactionalMapProxy<K, V>
 
     @Override
     @SuppressWarnings("unchecked")
-    public Collection<V> values(Predicate predicate) {
+    public Collection<V> values(Predicate<K, V> predicate) {
         checkTransactionState();
         checkNotNull(predicate, "Predicate can not be null!");
         checkNotInstanceOf(PagingPredicate.class, predicate,

```

**Review by srknzl — APPROVED** [review]

**Comment by srknzl** [comment]
run-lab-run

**Comment by aditya-32** [comment]
fixed check style error  , can you please run the pr builder once more?

**Comment by srknzl** [comment]
run-lab-run

**Comment by ahmetmircik** [comment]
@aditya-32 What was the rationale of this change? These are internal classes and type checking can stop working of it for some mechanisms i suppose.

**Review by ahmetmircik — CHANGES_REQUESTED** [review]
added question

**Comment by aditya-32** [comment]
Hi @ahmetmircik  can you please provide in some instance where we you see that can cause problem?

**Comment by aditya-32** [comment]
it was an open issue which i tried to fix by making the class generic

**Comment by aditya-32** [comment]
@srknzl @ahmetmircik  any update on this, Are we moving forward with this PR?

File: hazelcast/src/main/java/com/hazelcast/map/impl/tx/TransactionalMapProxy.java [hunk 2]
```diff
@@ -105,47 +105,47 @@ public boolean isEmpty() {
     }
 
     @Override
-    public Object get(Object key) {
+    public V get(Object key) {
         return get(key, false);
     }
 
-    public Object get(Object key, boolean skipNearCacheLookup) {
+    public V get(Object key, boolean skipNearCacheLookup) {
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/tx/TransactionalMapProxy.java:112]: let's keep this as Object return type, it serves for both client and member proxies and for client we sent serialized form.

**Comment by ahmetmircik** [comment]
@aditya-32 server side map-proxy is used for multiple purposes, returning both serialized and de-serialized is legitimate types for it. We shouldn't do generic conversion on this part. Given this, i'm sorry that we need to close this PR.

**Comment by aditya-32** [comment]
@ahmetmircik if there is issue in making the class generic sure we can close it 
just wanted to know like how making it generic will cause issue in serialisation and de-serialisation.

**Comment by ahmetmircik** [comment]
@aditya-32 thx for the work.

**Pull request closed without merging**